pkg/skaffold/kubernetes/debugging/container_manager.go

```diff
@@ -0,0 +1,151 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package debugging
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/debug"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/event"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes"
+)
+
+var (
+	// For testing
+	aggregatePodWatcher = kubernetes.AggregatePodWatcher
+
+	notifyDebuggingContainerStarted    = event.DebuggingContainerStarted
+	notifyDebuggingContainerTerminated = event.DebuggingContainerTerminated
+)
+
+type ContainerManager struct {
+	output      io.Writer
+	cli         *kubectl.CLI
+	podSelector kubernetes.PodSelector
+	namespaces  []string
+	active      map[string]string // set of containers that have been notified
+	aggregate   chan watch.Event
+}
+
+func NewContainerManager(out io.Writer, cli *kubectl.CLI, podSelector kubernetes.PodSelector, namespaces []string) *ContainerManager {
+	// Create the channel here as Stop() may be called before Start() when a build fails, thus
+	// avoiding the possibility of closing a nil channel. Channels are cheap.
+	return &ContainerManager{output: out, cli: cli, podSelector: podSelector, namespaces: namespaces, active: map[string]string{}, aggregate: make(chan watch.Event)}
+}
+
+func (d *ContainerManager) Start(ctx context.Context) error {
+	if d == nil {
+		// debug mode probably not enabled
+		return nil
+	}
+	stopWatchers, err := aggregatePodWatcher(d.namespaces, d.aggregate)
+	if err != nil {
+		stopWatchers()
+		return errors.Wrap(err, "initializing debugging container watcher")
+	}
+
+	go func() {
+		defer stopWatchers()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case evt, ok := <-d.aggregate:
+				if !ok {
+					return
+				}
+				// If the event's type is "ERROR", warn and continue.
+				if evt.Type == watch.Error {
+					logrus.Warnf("got unexpected event of type %s", evt.Type)
+					continue
+				}
+				// Grab the pod from the event.
+				pod, ok := evt.Object.(*v1.Pod)
+				if !ok {
+					continue
+				}
+				// Notify only when a container is Running or Terminated (not Waiting)
+				// "ADDED" is never interesting since no containers are Running yet
+				// "MODIFIED" may now have containers in Running or Terminated
+				// "DELETED" if the pod is deleted
+				if evt.Type != watch.Modified && evt.Type != watch.Deleted {
+					continue
+				}
```

Comment:
What I observed when prototyping this functionality that I couldn't rely on the pod status.  Rather I had to check on pod MODIFIED and DELETED events and then look at and track the individual container status (see below in `checkPod()`).  I would get spurious debug events otherwise.